File: Packages/BioMedLit/Sources/BioMedLit/Transparency/Analysis/TrialComplianceAnalyzer.swift
// BMLibrarian Lite - Biomedical Literature Research Tool
// Copyright (C) 2024-2025 Dr Horst Herb
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program. If not, see <https://www.gnu.org/licenses/>.

import Foundation

/// Pure functions for analyzing clinical trial compliance.
///
/// All functions are stateless and can be safely called from any context.
/// This module provides functionality for checking trial results posting compliance
/// based on FDAAA 2007 regulations, detecting clinical trials from study titles,
/// extracting NCT IDs from text, and analyzing outcome switching.
///
/// Usage:
/// ```swift
/// let compliance = TrialComplianceAnalyzer.checkResultsCompliance(
///     trial: trialReg,
///     publicationDate: pubDate
/// )
///
/// let nctIds = TrialComplianceAnalyzer.extractNCTIds(from: articleText)
/// ```
public enum TrialComplianceAnalyzer {

    // MARK: - Outcome Word Overlap Threshold

    /// Minimum word overlap ratio for matching registered outcomes to reported outcomes.
    /// A ratio above this threshold indicates the outcome is likely reported.
    private static let outcomeWordOverlapThreshold: Double = 0.5

    // MARK: - Results Compliance

    /// Check if trial results posting is compliant with regulations.
    ///
    /// Based on FDAAA 2007 which requires results within 12 months of completion
    /// for applicable clinical trials.
    ///
    /// - Parameters:
    ///   - trial: Trial registration information.
    ///   - publicationDate: Publication date of the article (optional, unused in current
    ///     implementation but reserved for future late detection logic).
    /// - Returns: Compliance status indicating whether results are compliant, missing,
    ///   or unknown.
    public static func checkResultsCompliance(
        trial: TrialRegistration,
        publicationDate: Date?
    ) -> ResultsComplianceStatus {
        // Results are posted - generally compliant
        if trial.resultsPosted {
            // Could add late detection with results posting date if available
            return .compliant
        }

        // No results posted - check if they should be
        guard let completionDate = trial.completionDate else {
            return .unknown
        }

        // Calculate deadline (12 months from completion)
        let deadline = Calendar.current.date(
            byAdding: .day,
            value: TransparencyConstants.resultsComplianceDeadlineDays,
            to: completionDate
        ) ?? completionDate

        // Check if past deadline
        if Date() > deadline {
            return .missing
        }

        return .unknown
    }

    // MARK: - Trial Detection

    /// Check if a study title suggests it's a clinical trial.
    ///
    /// Examines the title for common clinical trial indicators such as "randomized",
    /// "RCT", "phase II", "trial", etc.
    ///
    /// - Parameter title: Study title to analyze.
    /// - Returns: True if clinical trial indicators found, false otherwise.
    public static func appearsToBeClinicalTrial(title: String?) -> Bool {
        guard let title = title else { return false }

        let titleLower = title.lowercased()
        return ClinicalTrialPatterns.trialKeywords.contains { keyword in
            titleLower.contains(keyword)
        }
    }

    /// Extract NCT IDs from text.
    ///
    /// Searches for ClinicalTrials.gov registration identifiers in the format
    /// NCT followed by 8 digits (e.g., NCT01234567).
    ///
    /// - Parameter text: Text to search for NCT IDs.
    /// - Returns: Array of NCT IDs found, preserving original case.
    public static func extractNCTIds(from text: String) -> [String] {
        return RegexHelper.findAll(
            pattern: ClinicalTrialPatterns.nctIdPattern,
            in: text
        )
    }

    // MARK: - Industry Sponsor Detection

    /// Check if trial is industry-sponsored based on sponsor class.
    ///
    /// Uses the sponsor class field from ClinicalTrials.gov to determine
    /// if the trial is primarily sponsored by industry.
    ///
    /// - Parameter sponsorClass: Sponsor class from ClinicalTrials.gov
    ///   (e.g., "INDUSTRY", "NIH", "OTHER").
    /// - Returns: True if the sponsor class indicates industry sponsorship.
    public static func isIndustrySponsor(_ sponsorClass: String?) -> Bool {
        sponsorClass?.uppercased() == "INDUSTRY"
    }

    // MARK: - Outcome Analysis

    /// Check for potential outcome switching.
    ///
    /// Compares registered outcomes with reported outcomes using word overlap
    /// analysis to detect potential discrepancies between what was registered
    /// and what was reported.
    ///
    /// Note: This is a simplified check using word overlap. Full analysis
    /// would require NLP techniques for semantic similarity matching.
    ///
    /// - Parameters:
    ///   - registered: Registered primary outcomes from trial registry.
    ///   - reported: Reported outcomes extracted from publication.
    /// - Returns: Tuple of (detected, details) where detected is true if
    ///   potential outcome switching was found, and details contains
    ///   human-readable descriptions of discrepancies.
    public static func checkOutcomeSwitching(
        registered: [String],
        reported: [String]
    ) -> (detected: Bool, details: [String]) {
        // Skip if no registered outcomes to compare against
        guard !registered.isEmpty else {
            return (false, [])
        }

        // Skip if no reported outcomes to compare
        guard !reported.isEmpty else {
            return (false, ["Unable to compare - no reported outcomes extracted"])
        }

        var details: [String] = []

        // Simplified check: look for major mismatches using word overlap
        // A proper implementation would use NLP/fuzzy matching
        let registeredLower = Set(registered.map { $0.lowercased() })
        let reportedLower = Set(reported.map { $0.lowercased() })

        // Check if any registered outcomes are completely absent from reported
        for outcome in registeredLower {
            let found = reportedLower.contains { reported in
                // Simple word overlap check
                let outcomeWords = Set(outcome.components(separatedBy: .whitespaces)
                    .filter { !$0.isEmpty })
                let reportedWords = Set(reported.components(separatedBy: .whitespaces)
                    .filter { !$0.isEmpty })

                guard !outcomeWords.isEmpty else { return false }

                let overlap = outcomeWords.intersection(reportedWords)
                return Double(overlap.count) / Double(outcomeWords.count) > outcomeWordOverlapThreshold
            }

            if !found {
                // Truncate long outcome descriptions for display
                let truncatedOutcome = outcome.prefix(50)
                let suffix = outcome.count > 50 ? "..." : ""
                details.append("Registered outcome may not be reported: \(truncatedOutcome)\(suffix)")
            }
        }

        return (!details.isEmpty, details)
    }

    // MARK: - Summary

    /// Generate human-readable compliance summary.
    ///
    /// Creates a summary string describing the trial registration status
    /// and results compliance for display in the UI.
    ///
    /// - Parameters:
    ///   - registrations: Trial registrations found for the study.
    ///   - compliance: Results compliance status.
    /// - Returns: Summary string suitable for UI display.
    public static func formatSummary(
        registrations: [TrialRegistration],
        compliance: ResultsComplianceStatus
    ) -> String {
        guard !registrations.isEmpty else {
            return "No trial registration found"
        }

        let regCount = registrations.count
        let resultsPosted = registrations.filter(\.resultsPosted).count

        var parts: [String] = []
        parts.append("\(regCount) trial registration\(regCount == 1 ? "" : "s") found")

        if resultsPosted > 0 {
            parts.append("\(resultsPosted) with results posted")
        }

        switch compliance {
        case .compliant:
            parts.append("Results compliant")
        case .late:
            parts.append("Results posted late")
        case .missing:
            parts.append("Results missing")
        case .notRequired, .unknown:
            break
        }

        return parts.joined(separator: "; ")
    }

    // MARK: - Missing Registration Warning

    /// Generate warning if study appears to be a clinical trial without registration.
    ///
    /// Checks if the study title suggests it's a clinical trial but no
    /// registration was found, which may indicate a transparency concern.
    ///
    /// - Parameters:
    ///   - title: Study title.
    ///   - registrations: Found trial registrations (empty if none found).
    /// - Returns: Warning message if applicable, nil otherwise.
    public static func checkMissingRegistration(
        title: String?,
        registrations: [TrialRegistration]
    ) -> String? {
        guard registrations.isEmpty,
              appearsToBeClinicalTrial(title: title) else {
            return nil
        }

        return "Clinical trial without detected registration"
    }
}

File: android/MedicalFactChecker/app/src/main/java/com/bmlibrarian/factchecker/domain/transparency/TrialComplianceAnalyzer.kt
package com.bmlibrarian.factchecker.domain.transparency

import java.time.Instant
import java.time.ZoneId

/**
 * The outcome of comparing registered with reported trial outcomes.
 *
 * @property detected True if a registered outcome appears to be missing from the report.
 * @property details Human-readable descriptions of each discrepancy.
 */
data class OutcomeSwitchingResult(val detected: Boolean, val details: List<String>)

/**
 * Pure functions analysing clinical-trial registration and compliance.
 *
 * Ported from the Swift `TrialComplianceAnalyzer` (BioMedLit): FDAAA results
 * compliance, trial detection from titles, NCT ID extraction and a simple
 * outcome-switching check.
 */
object TrialComplianceAnalyzer {

    /** Word-overlap ratio above which a reported outcome counts as matching a registered one. */
    private const val OUTCOME_WORD_OVERLAP_THRESHOLD: Double = 0.5

    /** The sponsor class ClinicalTrials.gov reports for an industry sponsor. */
    private const val INDUSTRY_SPONSOR_CLASS: String = "INDUSTRY"

    /** Appended to an outcome description shortened for display. */
    private const val ELLIPSIS: String = "..."

    /**
     * Whether a trial's results posting complies with FDAAA 2007 (results within
     * 12 months of completion).
     *
     * Posted results are compliant. Unposted results are missing once the
     * deadline — completion plus [TransparencyConstants.RESULTS_COMPLIANCE_DEADLINE_DAYS]
     * calendar days in the device's time zone, as Swift's `Calendar.current`
     * computes it — has passed, and unknown before that or without a completion date.
     *
     * @param trial The trial registration.
     * @param publicationDate The article's publication date (unused, reserved for late detection, as in Swift).
     * @param now The current time; a parameter only so tests can fix it.
     * @return The compliance status.
     */
    @Suppress("UNUSED_PARAMETER")
    fun checkResultsCompliance(
        trial: TrialRegistration,
        publicationDate: Instant?,
        now: Instant = Instant.now(),
    ): ResultsComplianceStatus {
        if (trial.resultsPosted) return ResultsComplianceStatus.COMPLIANT

        val completionDate = trial.completionDate ?: return ResultsComplianceStatus.UNKNOWN
        val deadline = completionDate.atZone(ZoneId.systemDefault())
            .plusDays(TransparencyConstants.RESULTS_COMPLIANCE_DEADLINE_DAYS.toLong())
            .toInstant()

        return if (now.isAfter(deadline)) ResultsComplianceStatus.MISSING else ResultsComplianceStatus.UNKNOWN
    }

    /**
     * Whether a title suggests a clinical trial ("randomized", "RCT", "phase II", "trial", ...).
     *
     * @param title The study title, or null.
     * @return True if any [ClinicalTrialPatterns.trialKeywords] occurs in the lowercased title.
     */
    fun appearsToBeClinicalTrial(title: String?): Boolean {
        if (title == null) return false
        val titleLower = title.lowercase()
        return ClinicalTrialPatterns.trialKeywords.any { titleLower.contains(it) }
    }

    /**
     * Extract ClinicalTrials.gov NCT IDs (NCT + 8 digits) from text.
     *
     * @param text Text to search.
     * @return The IDs as they appear in [text], in order.
     */
    fun extractNCTIds(text: String): List<String> =
        TransparencyRegex.findAll(ClinicalTrialPatterns.NCT_ID_PATTERN, text)

    /**
     * Whether a ClinicalTrials.gov sponsor class denotes industry.
     *
     * @param sponsorClass The sponsor class (e.g. "INDUSTRY", "NIH", "OTHER"), or null.
     * @return True for "INDUSTRY", in any case.
     */
    fun isIndustrySponsor(sponsorClass: String?): Boolean =
        sponsorClass?.uppercase() == INDUSTRY_SPONSOR_CLASS

    /**
     * Check for potential outcome switching by word overlap.
     *
     * A registered outcome counts as reported when more than half of its words
     * appear in some reported outcome. A simplified check; semantic matching
     * would need NLP.
     *
     * @param registered Registered outcomes.
     * @param reported Outcomes extracted from the publication.
     * @return Whether switching was detected, with one detail per unmatched registered outcome.
     */
    fun checkOutcomeSwitching(registered: List<String>, reported: List<String>): OutcomeSwitchingResult {
        if (registered.isEmpty()) return OutcomeSwitchingResult(detected = false, details = emptyList())
        if (reported.isEmpty()) {
            return OutcomeSwitchingResult(
                detected = false,
                details = listOf("Unable to compare - no reported outcomes extracted"),
            )
        }

        val registeredLower = registered.map { it.lowercase() }.toCollection(LinkedHashSet())
        val reportedLower = reported.map { it.lowercase() }.toCollection(LinkedHashSet())
        val details = mutableListOf<String>()

        for (outcome in registeredLower) {
            val outcomeWords = words(outcome)
            val found = reportedLower.any { candidate ->
                if (outcomeWords.isEmpty()) return@any false
                val overlap = outcomeWords.intersect(words(candidate))
                overlap.size.toDouble() / outcomeWords.size > OUTCOME_WORD_OVERLAP_THRESHOLD
            }
            if (!found) {
                val maxLength = TransparencyConstants.MAX_OUTCOME_DESCRIPTION_LENGTH
                val suffix = if (outcome.length > maxLength) ELLIPSIS else ""
                details.add("Registered outcome may not be reported: ${outcome.take(maxLength)}$suffix")
            }
        }

        return OutcomeSwitchingResult(detected = details.isNotEmpty(), details = details)
    }

    /**
     * The distinct words of [text], split on spaces and tabs as Swift's
     * `CharacterSet.whitespaces` does (line breaks are not separators).
     */
    private fun words(text: String): Set<String> =
        text.splitWhere { it == '\t' || Character.getType(it) == Character.SPACE_SEPARATOR.toInt() }
            .filter { it.isNotEmpty() }
            .toSet()

    /** Split on every character matching [isSeparator]. */
    private inline fun String.splitWhere(isSeparator: (Char) -> Boolean): List<String> {
        val parts = mutableListOf<String>()
        val current = StringBuilder()
        for (character in this) {
            if (isSeparator(character)) {
                parts.add(current.toString())
                current.clear()
            } else {
                current.append(character)
            }
        }
        parts.add(current.toString())
        return parts
    }

    /**
     * A one-line summary of trial registration and results compliance.
     *
     * @param registrations Registrations found for the study.
     * @param compliance Results compliance status.
     * @return The summary.
     */
    fun formatSummary(registrations: List<TrialRegistration>, compliance: ResultsComplianceStatus): String {
        if (registrations.isEmpty()) return "No trial registration found"

        val regCount = registrations.size
        val resultsPosted = registrations.count { it.resultsPosted }
        val parts = mutableListOf("$regCount trial registration${if (regCount == 1) "" else "s"} found")

        if (resultsPosted > 0) parts.add("$resultsPosted with results posted")

        when (compliance) {
            ResultsComplianceStatus.COMPLIANT -> parts.add("Results compliant")
            ResultsComplianceStatus.LATE -> parts.add("Results posted late")
            ResultsComplianceStatus.MISSING -> parts.add("Results missing")
            ResultsComplianceStatus.NOT_REQUIRED, ResultsComplianceStatus.UNKNOWN -> Unit
        }

        return parts.joinToString("; ")
    }

    /**
     * A warning when a study looks like a clinical trial but no registration was found.
     *
     * Only a registry's answer can say a registration is missing: with no trial ID to look
     * up, or no answer, the list is empty for a reason that is not the study's, so nothing is
     * said (mirrors Python's `trial_registration_assessed` gate).
     *
     * @param title The study title.
     * @param registrations Registrations found (empty if none).
     * @param registrationAssessed Whether ClinicalTrials.gov answered for every trial the
     *   article cites.
     * @return [RiskIndicatorStrings.MISSING_TRIAL_REGISTRATION], or null.
     */
    fun checkMissingRegistration(
        title: String?,
        registrations: List<TrialRegistration>,
        registrationAssessed: Boolean,
    ): String? {
        if (!registrationAssessed || registrations.isNotEmpty() || !appearsToBeClinicalTrial(title)) return null
        return RiskIndicatorStrings.MISSING_TRIAL_REGISTRATION
    }

    /**
     * Warning for a trial whose registry lookup failed (network, server or HTTP error): its
     * registration was not checked, not found missing.
     *
     * @param nctId The trial's NCT ID.
     * @return The warning, worded as the Python reference words it.
     */
    fun registryUnreachableWarning(nctId: String): String =
        "Could not reach ClinicalTrials.gov for trial $nctId, so its registration could " +
            "not be checked. Absence of a registration is not evidence the study is unregistered."

    /**
     * Warning for a trial ClinicalTrials.gov answered it holds no record of.
     *
     * @param nctId The trial's NCT ID.
     * @return The warning — a finding about the study, not about the lookup.
     */
    fun registryHasNoRecordWarning(nctId: String): String =
        "ClinicalTrials.gov holds no record of trial $nctId, which the article cites as its registration."

    /**
     * Warning for a registry record that arrived but could not be read.
     *
     * @param nctId The trial's NCT ID.
     * @return The warning; an unread record is not an absent one.
     */
    fun unreadableRegistryRecordWarning(nctId: String): String =
        "ClinicalTrials.gov returned a record for trial $nctId that could not be read, so its " +
            "registration could not be checked."
}
